keps/sig-apps/20190318-Graduate-CronJob-to-Stable.md

```diff
@@ -0,0 +1,579 @@
+		return
+	}
+    for i := 0; i < workers; i++ {
+        go wait.Until(cjc.worker, time.Second, stopCh)
+    }
+...
+}
+
+func (cjc *CronJobController) func worker() {
+    for cjc.processNextWorkItem() {}
+}
+
+func (cjc *CronJobController) func processNextWorkItem() {
+	key, quit := cjc.queue.Get()
+	if quit {
+		return false
+	}
+	defer cjc.queue.Done(key)
+	if err := cjc.sync(key.(string)); err != nil {
+		utilruntime.HandleError(fmt.Errorf("Error syncing CronJobController %v, requeuing: %v", key.(string), err))
+		cjc.queue.AddRateLimited(key)
+	} else {
+		cjc.queue.Forget(key)
+	}
+	return true
+}
+
+func (cjc *CronJobController)  sync(cronJobKey) {
+	childrenJobs := cjc.jobLister.GetJobsForCronJob(cronJobKey)
+	// reuse/refactor the existing syncOne function
+}
+
+
+```
+
+
+##### Multiple workers
+We also propose to have multiple workers controller by a flag similar to [statefulset controller](https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-controller-manager/app/apps.go#L65). The default would be set to 5 similar to [statefulset](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/statefulset/config/v1alpha1/defaults.go#L34)
+
+```
+// Controller code
+
+func (cjc *CronJobController)  Run() {
+	...
+    for i := 0; i < workers; i++ {
+        go wait.Until(cjc.worker, time.Second, stopCh)
+    }
+...
+}
+
+```
+
+##### Handling Cron aspect
+To detect which CronJob has met its schedule and need to create Jobs we need to implement a timer component. These are the possible options for implementing the timer:
+
+| algorithm  | how it works | notes |
+|:----------|:----------|:-------|
+| Unordered timer list | Periodic Sweep from cache | Slower and similar to existing implementation. But improved because we sweep fom the cache instead of API server | 
+|Ordered timer list| Maintain ordered list of Cronjob keys and next time of expiry. Keep starting a timer with the earliest expiry. | Efficient. Reinsertion to list takes O(n) |
+|Timer trees| Instead of ordered list use a sorted binary tree. | More efficient. Insertion is O(log n) |
+|Heap based timer|A variant of ordered timer list where heap is used to store the next expiry time | Efficient compared to ordered list. Bookkeeping and insertion is O(log n). |
+|Simple Timing wheels| circular buffer of MaxTimeOut slots. List of expiring timers at each slot. | Works for small bounded  MaxTimeOut which is not our case. Insertion and removal is O(1) via indexing |
+|Hashed Wheel| Hash expiring time and insert in a hash table with linked list at each index | Bookkeeping is O(1) and worst case insertion is O(n) |
+|Hierarchical Wheel| multiple timer wheels for different resolutions (Seconds, minutes, hours, days). When seconds rolls over we grab the next minutes timers and recreate the seconds wheel. similarly for minutes and hours. | Sharding at different hierarchy levels improves insertion and bookkeeping performance. |
```

Comment:
I would suggest moving this whole table to considered alternatives section (together with further reading).
This allows people interested in just looking how we do thing have a shorter doc to read (and those who wants more details dig deeper into alternatives section).

Reply:
ack. did so.